### ISSUE TITLE
Incredibly slow upload speeds

### ISSUE THREAD

**Issue by DanielMarquard** [s3fs-fuse/s3fs-fuse#798]
I'm using s3fs for one of my use cases after I encountered a problem with goofys (which is super fast), that won't be possible for me to work around. So I'm curious, is 2 MB/s upload to an internal S3 endpoint nominal for s3fs? It just doesn't sound right.

My mount is simple:
```
s3fs bucket-name /mnt/dir -o uid=$(id -u user) -o gid=$(id -g group) -o allow_other -o iam_role="$(iam_role)"
```

Just curious if anyone else has this problem? I mean, I'm grateful that it's at least working, but at 2 MB/s?

**Comment by ggtakec** [comment]
You can try to use latest version(1.86) or master branch code.
It is tuned performance.
And I recommend you should use use_cache and try to specify multipart_size option(@gaul said, too).
Thanks in advance for your help.

**Comment by pwulff** [comment]
I get the same kind of performance on my private S3 with the master branch.

**Comment by gaul** [comment]
It would help to explain exactly which operations exhibit poor performance and compare them with another tool like AWS CLI.  Also please quantify this with file size and MB/s.

**Comment by pwulff** [comment]
I have a data set of 2000 binary files of 1.000.000 bytes created with the command line:
`for i in {1..2001}; do dd if=/dev/urandom bs=1000000 count=1 of=file$i; done`
I copy them with the command line:
`rsync --info=progress2 /src/dataset/* /dst/s3fs-fuse-mountpoint/folder`
And i got between 1.5 and 2.5 MB/s.

I have made the same transfert with a Windows client and I got 12.5 MB/s.
My S3 endpoint is a local private cloud and we have a 1Gb network.

**Comment by pwulff** [comment]
I have made a test with `dbglevel=info -f -o curldbg` as requested by @gaul and I got a lot of 404.

`[INF]       curl.cpp:RequestPerform(2250): HTTP response code 404 was returned, returning ENOENT`

**Comment by pwulff** [comment]
After more test I've found this:
`[WAN] curl.cpp:ResetHandle(1855): The CURLOPT_SSL_ENABLE_ALPN option could not be unset. S3 server does not support ALPN, then this option should be disabled to maximize performance. you need to use libcurl 7.36.0 or later.
[WAN] curl.cpp:ResetHandle(1858): The S3FS_CURLOPT_KEEP_SENDING_ON_ERROR option could not be set. For maximize performance you need to enable this option and you should use libcurl 7.51.0 or later.`

I will try to update my RHEL 7.6 with a more recent libcurl to see if that change anything.

**Comment by pwulff** [comment]
No change of speed with curl 7.64.1 installed.

**Comment by ggtakec** [comment]
@pwulff 
If objects in your bucket are only files/directories created by s3fs, you can try to specify the notsup_compat_dir option.
I think that using this option will reduce 404 errors.

s3fs performs checks to recognize directories of paths to objects created by other tools, and it generates 404 errors.
This option does not check these another type objects, then it not makes 404 error.

Thanks in advance for your assistance.

**Comment by pwulff** [comment]
I've profiled s3fs during my testing and following callgrind 70% of the time is consumed by EVP_DigestUpdate(). To mount the bucket, upload 300 files of 1.000.000 of bytes and unmount the bucket this method is called more than 800.000 times. I suppose that the SHA256 hash overhead is not so painful for big files.

**Comment by ggtakec** [comment]
@pwulff 
EVP_DigestUpdate is called to create a signature to access S3.
And when you mount, many HEAD requests are sent to check the existence of directories and files in order to upload files.
I think that the number of requests has increased because of this confirmation, and the number of calls to EVP_DigestUpdate is also increasing.

Could you try the option (max_stat_cache_size, stat_cache_expire, stat_cache_interval_expire, enable_noobj_cache) to cache the file stat info once confirmed?
Thanks in advance for your assistance.

**Comment by joshglenn** [comment]
I'm also having this problem... on Mac osx mojave, copying 19 files for a total of 2.0 MB takes about 10 minutes. Slow as dial-up :-)

**Comment by nikt12** [comment]
@ggtakec Hello, could you advice, what options should be to upload large files very fast?
I am using 1.84 version and try to upload just one file to S3 compatible storage.
Next options were used: "-o use_path_request_style -o enable_content_md5 -o allow_other -o nonempty -o kernel_cache -o max_stat_cache_size=1000000 -o parallel_count=200 -o multipart_size=100 -o singlepart_copy_limit=200 -o multireq_max=30 -o max_background=1000".
With this I could reach only near 60 MB/S uploading speed.
I'm under assumption that multipart uploading was not enabled somehow (I had 35GB file, and I think that several threads have to upload this, but with htop CPU usage for s3fs daemon was not higher than 100%).

**Comment by gaul** [comment]
Could you test with the latest version 1.85?  It includes several optimizations which should improve your performance.

**Comment by gaul** [comment]
Please reopen if symptoms persist.

**Comment by neilpalima** [comment]
@gaul, I think I'm experiencing this issue. Is there a way how to verify this? On an aws ec2 instance, we are writing a file with ~70mb but it takes like 40mins-1hour(or more) to be uploaded in our s3 bucket. The commit hash we are using is `005a684600f83a784c526ada078ac807e5e19633` - `Fix typos`.

**Comment by drzraf** [comment]
I found that on openstack, in a preliminary step, for each file to upload using rsync, 4 requests are made sequentially.
- `/2018-09/x.mp4_$folder$`
- `/2018-09/x.mp4/`
- `/2018-09/x.mp4`
- `delimiter=/&max-keys=2&prefix=2018-09/x.mp4/`
And I doubt files are processed in parallel.

As such, before the upload of 500 files even starts, 2000k requests are issued. This is at least suboptimal.

**Comment by gaul** [comment]
@neilpalima s3fs should sustain 100+ MBytes/s for sequential reads and writes.  Please open a new issue with your symptoms and benchmark setup if you experience poor performance.

@drzraf s3fs does not have optimal metadata performance and does some checks sequentially.  #927 discusses changing these defaults.  Note that this should not dramatically affect performance for large files.

**Comment by hopeseekr** [comment]
`cp -rvf` works fine and fast. `rsync -avW --progress --inplace --size-only` seems to get frozen at `sending incremental file list`... all it does is create all of the directories.

**Comment by skepticalwaves** [comment]
I'm having the same experience as @hopeseekr, `cp` works, but rsync hangs forever after creating directories.
